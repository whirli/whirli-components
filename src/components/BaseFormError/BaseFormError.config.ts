// @ts-ignore
import { ComponentStyles, ComponentProps } from '@whirli-components/@types/components';

// Local config
import config from '@whirli-local/components/BaseFormError/BaseFormError.config';

// Package config
export const ConfigStyles: ComponentStyles = {
  ...config.styles,
};

export const ConfigProps: ComponentProps = {
  ...config.props,
};
