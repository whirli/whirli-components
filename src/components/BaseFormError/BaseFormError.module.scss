@import '@whirli-components/styles/Breakpoints.module.scss';

// default styles
.error {}
// breakpoint styles
$packageBreakpointStyles: ();