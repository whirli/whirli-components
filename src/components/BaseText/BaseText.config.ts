// @ts-ignore
import { PropType } from '@composition';
import * as PropHelpers from '@whirli-components/helpers/props';
import { ComponentStyles, ComponentProps } from '@whirli-components/@types/components';

// Local config
import config from '@whirli-local/components/BaseText/BaseText.config';

// Package config
import { PropKeys, PropValues } from './BaseText.constants';
import type {
  PropColor,
  PropSize,
  PropTag,
  PropFont,
  PropTransform,
  PropLetterSpacing,
  PropLineHeight,
  PropAlign,
} from './BaseText.types';

export const ConfigStyles: ComponentStyles = {
  ...config.styles,
  [PropKeys.COLOR]: {
    generateBreakpoints: true,
    classes: {
      default: 'color-inherit',
      ...config.styles[PropKeys.COLOR]?.classes,
    } as Record<PropColor, string>,
  },
  [PropKeys.SIZE]: {
    generateBreakpoints: true,
    classes: {
      '3xs': 'size-3xs',
      '2xs': 'size-2xs',
      xs: 'size-xs',
      sm: 'size-sm',
      default: 'size-default',
      md: 'size-md',
      lg: 'size-lg',
      xl: 'size-xl',
      '2xl': 'size-2xl',
      '3xl': 'size-3xl',
      '4xl': 'size-4xl',
      '5xl': 'size-5xl',
      ...config.styles[PropKeys.SIZE]?.classes,
    } as Record<PropSize, string>,
  },
  [PropKeys.FONT]: {
    generateBreakpoints: true,
    classes: {
      default: 'font-sans',
      bold: 'font-bold',
      mono: 'font-mono',
      ...config.styles[PropKeys.FONT]?.classes,
    } as Record<PropFont, string>,
  },
  [PropKeys.TRANSFORM]: {
    classes: {
      default: 'transform-default',
      uppercase: 'transform-uppercase',
      ...config.styles[PropKeys.TRANSFORM]?.classes,
    } as Record<PropTransform, string>,
  },
  [PropKeys.LETTER_SPACING]: {
    generateBreakpoints: true,
    classes: {
      default: 'letter-spacing-default',
      md: 'letter-spacing-md',
      lg: 'letter-spacing-lg',
      xl: 'letter-spacing-xl',
      ...config.styles[PropKeys.LETTER_SPACING]?.classes,
    } as Record<PropLetterSpacing, string>,
  },
  [PropKeys.LINE_HEIGHT]: {
    generateBreakpoints: true,
    classes: {
      '3xs': 'line-height-3xs',
      '2xs': 'line-height-2xs',
      xs: 'line-height-xs',
      sm: 'line-height-sm',
      default: 'line-height-default',
      md: 'line-height-md',
      lg: 'line-height-lg',
      xl: 'line-height-xl',
      '2xl': 'line-height-2xl',
      '3xl': 'line-height-3xl',
      '4xl': 'line-height-4xl',
      '5xl': 'line-height-5xl',
      ...config.styles[PropKeys.LINE_HEIGHT]?.classes,
    } as Record<PropLineHeight, string>,
  },
  [PropKeys.ALIGN]: {
    generateBreakpoints: true,
    classes: {
      default: 'align-default',
      center: 'align-center',
      right: 'align-right',
      ...config.styles[PropKeys.ALIGN]?.classes,
    } as Record<PropAlign, string>,
  },
};

export const ConfigProps: ComponentProps = {
  [PropKeys.COLOR]: {
    required: false,
    type: [String, Object] as PropType<PropColor>,
    default: 'default',
    ...PropHelpers.breakpointStylePropValidation(ConfigStyles, PropKeys.COLOR),
  },
  [PropKeys.TAG]: {
    required: false,
    type: String as PropType<PropTag>,
    validator: (value: PropTag): boolean => PropValues[PropKeys.TAG].includes(value),
    default: 'span',
  },
  [PropKeys.SIZE]: {
    required: false,
    type: [String, Object] as PropType<PropSize>,
    default: 'default',
    ...PropHelpers.breakpointStylePropValidation(ConfigStyles, PropKeys.SIZE),
  },
  [PropKeys.FONT]: {
    required: false,
    type: [String, Object] as PropType<PropFont>,
    default: 'default',
    ...PropHelpers.breakpointStylePropValidation(ConfigStyles, PropKeys.FONT),
  },
  [PropKeys.TRANSFORM]: {
    required: false,
    type: String as PropType<PropTransform>,
    default: 'default',
    ...PropHelpers.stylePropValidation(ConfigStyles, PropKeys.TRANSFORM),
  },
  [PropKeys.LETTER_SPACING]: {
    required: false,
    type: [String, Object] as PropType<PropLetterSpacing>,
    default: 'default',
    ...PropHelpers.breakpointStylePropValidation(ConfigStyles, PropKeys.LETTER_SPACING),
  },
  [PropKeys.LINE_HEIGHT]: {
    required: false,
    type: [String, Object] as PropType<PropLineHeight>,
    default: 'default',
    ...PropHelpers.breakpointStylePropValidation(ConfigStyles, PropKeys.LINE_HEIGHT),
  },
  [PropKeys.ALIGN]: {
    required: false,
    type: [String, Object] as PropType<PropAlign>,
    default: 'default',
    ...PropHelpers.breakpointStylePropValidation(ConfigStyles, PropKeys.ALIGN),
  },
  ...config.props,
};
