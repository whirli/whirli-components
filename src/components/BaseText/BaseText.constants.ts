export const PropKeys = {
  COLOR: 'color',
  TAG: 'tag',
  SIZE: 'size',
  FONT: 'font',
  TRANSFORM: 'transform',
  LETTER_SPACING: 'letterSpacing',
  LINE_HEIGHT: 'lineHeight',
  ALIGN: 'align',
};

export const PropValues = {
  [PropKeys.COLOR]: ['inherit'],
  [PropKeys.TAG]: [
    'a',
    'nuxt-link',
    'router-link',
    'address',
    'time',
    'span',
    'strong',
    'div',
    'p',
    'h1',
    'h2',
    'h3',
    'h4',
    'h5',
    'h6',
    'address',
    'li',
    'blockquote',
    'dt',
    'dd',
    'figcaption',
  ],
  [PropKeys.SIZE]: ['3xs', '2xs', 'xs', 'sm', 'base', 'md', 'lg', 'xl', '2xl', '3xl', '4xl', '5xl'],
  [PropKeys.FONT]: ['default', 'bold', 'mono'],
  [PropKeys.TRANSFORM]: ['default', 'uppercase'],
  [PropKeys.LETTER_SPACING]: ['default', 'md', 'lg', 'xl'],
  [PropKeys.LINE_HEIGHT]: ['3xs', '2xs', 'xs', 'sm', 'base', 'md', 'lg', 'xl', '2xl', '3xl', '4xl', '5xl'],
  [PropKeys.ALIGN]: ['default', 'center', 'right'],
};
