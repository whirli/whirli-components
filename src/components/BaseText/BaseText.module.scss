@import '../../styles/Breakpoints.scss';
@import '../../styles/Helpers.scss';

@font-face {
    font-family: "Maax";
    src: url('~@whirli-components/src/assets/fonts/Maax.woff2') format("woff2");
}

@font-face {
    font-family: "MaaxBold";
    src: url('~@whirli-components/src/assets/fonts/MaaxBold.woff2') format("woff2");
}

@font-face {
    font-family: "MaaxMono";
    src: url('~@whirli-components/src/assets/fonts/MaaxMonoRegular.woff2') format("woff2");
}

// default styles
.text {}

.transform-default {
    text-transform: none;
}

.transform-uppercase {
    text-transform: uppercase;
}

.font-sans {
    font-family: 'Maax', Tahoma, Geneva, "Helvetica Neue", Helvetica, Arial, sans-serif;
}

.font-bold {
    font-family: 'MaaxBold', Tahoma, Geneva, "Helvetica Neue", Helvetica, Arial, sans-serif;
    font-weight: bold;
}

.font-mono {
    font-family: 'MaaxMono', "Courier New", Courier, "Lucida Sans Typewriter", "Lucida Typewriter", monospace;
}


// breakpoint styles
$packageBreakpointStyles: (
    color-inherit: (
        color: inherit,
    ),
    size-3xs: (
        font-size: toRems(10),
    ),
    size-2xs: (
        font-size: toRems(12),
    ),
    size-xs: (
        font-size: toRems(13),
    ),
    size-sm: (
        font-size: toRems(14),
    ),
    size-default: (
        font-size: toRems(16),
    ),
    size-md: (
        font-size: toRems(18),
    ),
    size-lg: (
        font-size: toRems(20),
    ),
    size-xl: (
        font-size: toRems(24),
    ),
    size-2xl: (
        font-size: toRems(28),
    ),
    size-3xl: (
        font-size: toRems(32),
    ),
    size-4xl: (
        font-size: toRems(40),
    ),
    size-5xl: (
        font-size: toRems(48),
    ),
    line-height-3xs: (
        font-size: toRems(16),
    ),
    line-height-2xs: (
        font-size: toRems(18),
    ),
    line-height-xs: (
        font-size: toRems(20),
    ),
    line-height-sm: (
        font-size: toRems(22),
    ),
    line-height-default: (
        font-size: toRems(24),
    ),
    line-height-md: (
        font-size: toRems(28),
    ),
    line-height-lg: (
        font-size: toRems(32),
    ),
    line-height-xl: (
        font-size: toRems(36),
    ),
    line-height-2xl: (
        font-size: toRems(40),
    ),
    line-height-3xl: (
        font-size: toRems(42),
    ),
    line-height-4xl: (
        font-size: toRems(48),
    ),
    line-height-5xl: (
        font-size: toRems(56),
    ),
    letter-spacing-default: (
        font-size: toRems(0),
    ),
    letter-spacing-md: (
        font-size: toRems(1),
    ),
    letter-spacing-lg: (
        font-size: toRems(1.2),
    ),
    letter-spacing-xl: (
        font-size: toRems(1.5),
    ),
    align-default: (
        text-align: left,
    ),
    align-center: (
        text-align: center,
    ),
    align-right: (
        text-align: right,
    )
);