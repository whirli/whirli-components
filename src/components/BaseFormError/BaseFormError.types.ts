import { PropKeys, PropValues } from './BaseFormError.constants';
