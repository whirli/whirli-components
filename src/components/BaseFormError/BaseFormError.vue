<template>
  <BaseText v-bind="$attrs" :class="classes" v-if="$slots.default"><slot /></BaseText>
</template>

<script setup lang="ts">
// Styles
// @ts-ignore
import styles from '@whirli-local/components/BaseFormError/BaseFormError.module.scss?module';

// Data
import { ConfigStyles, ConfigProps } from './BaseFormError.config';

// Types
import { ComponentStyles as ComponentStylesInterface } from '@whirli-components/@types/components';
import { Props } from '@whirli-components/@types/props';

const ComponentStyles: ComponentStylesInterface = ConfigStyles;

// @ts-ignore
const props: Props = defineProps(ConfigProps);

// Classes
import useClasses from '@whirli-components/@use/class';
import BaseText from '@whirli-components/components/BaseText/BaseText.vue';
const { makeClasses } = useClasses();
const classes = [styles.error, ...makeClasses(ComponentStyles, ConfigProps, props, styles)];
</script>
