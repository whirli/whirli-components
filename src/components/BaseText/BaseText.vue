<template>
  <component :is="props.tag" :class="classes"><slot /></component>
</template>

<script setup lang="ts">
// Vue
// Styles
import styles from '@whirli-local/components/BaseText/BaseText.module.scss?module';

// Data
import { ConfigStyles, ConfigProps } from './BaseText.config';

const ComponentStyles = ConfigStyles;

const props = defineProps(ConfigProps);

// Classes
import useClasses from '@whirli-components/@use/class';
const { makeClasses } = useClasses();
const classes = [styles.text, ...makeClasses(ComponentStyles, props, styles)];
</script>
