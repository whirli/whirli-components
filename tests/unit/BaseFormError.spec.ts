import { mount } from '@vue/test-utils';
import BaseFormError from '@/components/BaseFormError/BaseFormError.vue';
import BaseText from '@/components/BaseText/BaseText.vue';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
let wrapper: any;
const text = 'Hello World';

describe('BaseFormLabel.vue - With Slot', () => {
  beforeEach(() => {
    wrapper = mount(BaseFormError, {
      slots: {
        default: text,
      },
    });
  });

  it('renders the component', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('displays BaseText', () => {
    expect(wrapper.findComponent(BaseText).exists()).toBeTruthy();
  });

  it('displays the correct text', () => {
    expect(wrapper.text()).toMatch(text);
  });
});

describe('BaseFormLabel.vue - Without Slot', () => {
  beforeEach(() => {
    wrapper = mount(BaseFormError, {});
  });

  it('renders the component', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('displays BaseText', () => {
    expect(wrapper.findComponent(BaseText).exists()).toBeFalsy();
  });
});
