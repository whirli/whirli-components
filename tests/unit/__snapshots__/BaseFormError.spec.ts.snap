// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseFormLabel.vue - With Slot renders the component 1`] = `
<span
  class=""
>
  
  
  Hello World
  
  
</span>
`;

exports[`BaseFormLabel.vue - Without Slot renders the component 1`] = `<!--v-if-->`;
